### PR TITLE
Support HQT on VLLM

### PR DESCRIPTION
**Pull request by nirda7**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -286,6 +286,8 @@ def forward(
         else:
             hidden_states = self.get_input_embeddings(input_ids)
         residual = None
+        import habana_frameworks.torch as htorch
```
Comment (kzawora-intel) [inline vllm/model_executor/models/llama.py:289]: this will break execution on non-hpu accelerators

File: vllm/utils.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
     "half": torch.half,
     "bfloat16": torch.bfloat16,
     "float": torch.float,
-    "fp8": torch.uint8,
+    "fp8": torch.float8_e4m3fn,
```
Comment (kzawora-intel) [inline vllm/utils.py:34]: why is that needed? are we sure we're not breaking compatibility with this?

File: vllm/worker/habana_model_runner.py [hunk 3]
```diff
@@ -942,10 +956,16 @@ def log_warmup(self, phase, i, max_i, batch_size, seq_len):
...
         for i, (batch_size, seq_len) in enumerate(reversed(buckets)):
             mem_usage = 100.0 * HabanaMemoryProfiler.current_device_memory_usage() / HabanaMemoryProfiler.total_device_memory()
             self.log_warmup('Prompt' if is_prompt else 'Decode', i, len(buckets), batch_size, seq_len)
-            self.warmup_scenario(batch_size, seq_len, is_prompt, kv_caches)
+            try:
+                self.warmup_scenario(batch_size, seq_len, is_prompt, kv_caches)
+            except:
+                print(f"Failed on scenario {i+1}: batch_size={batch_size}, seq_len={seq_len}, is_prompt={is_prompt}")
+                counter += 1
+        print(f"Failed warm-up scenarios = {counter}")
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:968]: warmup should not fail, and if it does, we should not suppress those failures

File: vllm/worker/habana_model_runner.py [hunk 4]
```diff
@@ -971,11 +992,16 @@ def warmup_graphs(self, strategy, buckets, is_prompt, kv_caches, available_mem):
             self.graphed_buckets.add((batch_size, seq_len, is_prompt))
             self.log_warmup(phase, idx, num_candidates, batch_size, seq_len)
             with HabanaMemoryProfiler() as mem_prof:
-                self.warmup_scenario(batch_size, seq_len, is_prompt, kv_caches)
+                try:
+                    self.warmup_scenario(batch_size, seq_len, is_prompt, kv_caches)
+                except:
+                    print(f"Failed on graph scenario {idx}: batch_size={batch_size}, seq_len={seq_len}, is_prompt={is_prompt}")
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:998]: same comment as previously - warmup should not fail, and if it does, we should not suppress those failures

File: vllm/worker/habana_model_runner.py [hunk 5]
```diff
@@ -1008,6 +1034,19 @@ def warmup_model(self, kv_caches: List[torch.Tensor]) -> None:
...
         self.profiler.end()
 
+    def shutdown_hqt(self):
+        print('hqt shutdown')
+        if model_config := getattr(self, "model_config", None):
+            if getattr(model_config, "quantization", None) == 'hqt':
+                print('hqt shutdown start')
+                import habana_quantization_toolkit
+                if habana_quantization_toolkit is not None:
+                    habana_quantization_toolkit.finish_measurements(self.model.model)
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:1044]: self.model.model might not exist here

File: vllm/hpu/ops.py [hunk 6]
```diff
@@ -38,8 +38,8 @@ def fetch_from_cache(cache, blocks, permutations):
     return [cache.index_select(0, blocks[:, i]).permute(permutations) for i in range(blocks.size(1))]
 
 
-@hpu_utils.with_mark_steps
-def paged_attention_v1(query, key_cache, value_cache, head_mapping, scale, block_tables, context_lens, block_size, alibi_slopes, kv_cache_dtype=None) -> None:
+def paged_attention_v1(query, key_cache, value_cache, head_mapping, scale, block_tables, context_lens, block_size, alibi_slopes, kv_cache_dtype=None,
```
Comment (kzawora-intel) [inline vllm/hpu/ops.py:41]: why is @hpu_utils.with_mark_steps removed here?

Comment (bgoldberg-habana) [reply vllm/hpu/ops.py:41]: we want that all convert to\from hf8 will be in the same graph
so we remove this mark step and add one outside the transformer block align to ohf version.

File: vllm/attention/backends/habana_attn.py [hunk 7]
```diff
@@ -188,11 +196,9 @@ def forward(
...
             # If kv_cache is not provided, the new key and value tensors are
             # not cached. This happens during the initial memory profiling run.
-            HabanaPagedAttention.write_to_paged_cache(key, value, key_cache,
-                                                      value_cache,
-                                                      attn_metadata.slot_mapping,
-                                                      attn_metadata.kv_cache_dtype,
-                                                      attn_metadata.prefill_metadata is not None)
+            block_indices, block_offset = cache_ops.prepare_to_cache(key_cache, attn_metadata.slot_mapping)
+            key_cache = self.key_cache(key, key_cache, block_indices, block_offset)
+            value_cache = self.value_cache(value, value_cache, block_indices, block_offset)
```
Comment (kzawora-intel) [inline vllm/attention/backends/habana_attn.py:201]: @madamczykhabana can you check if this is fine? aren't we making HabanaPagedAttention.write_to_paged_cache useless here?

File: vllm/model_executor/models/llama.py (hunk 1 above)
Comment (nirda7) [reply vllm/model_executor/models/llama.py:289]: Added check for is_hpu.

File: vllm/utils.py (hunk 2 above)
Comment (nirda7) [reply vllm/utils.py:34]: added hf8 for float8_e4m3fn

File: vllm/worker/habana_model_runner.py (hunk 4 above)
Comment (nirda7) [reply vllm/worker/habana_model_runner.py:998]: Removed

File: vllm/worker/habana_model_runner.py (hunk 5 above)
Comment (nirda7) [reply vllm/worker/habana_model_runner.py:1044]: Why?
you wrap it with HpuModelAdapter anyway.

File: vllm/worker/habana_model_runner.py (hunk 3 above)
Comment (nirda7) [reply vllm/worker/habana_model_runner.py:968]: Removed

**Comment by madamczykhabana** [comment]
Relevant changes moved to #75

File: vllm/model_executor/layers/layernorm.py [hunk 8]
```diff
@@ -72,8 +72,8 @@ def forward(
             return x, residual
         if x.device.type == "hpu" and FusedRMSNorm:
             orig_dtype = x.dtype
```
Comment (dudilester) [inline vllm/model_executor/layers/layernorm.py:74]: orig_dtype not used, here and also in line 60
